fdbrpc/FlowTests.actor.cpp

```diff
@@ -1617,3 +1621,175 @@ TEST_CASE("/flow/flow/FlowMutex") {
+		}
+		stream.sendError(end_of_stream());
+		return Void();
+	});
+
+	state int n = 0;
+	int r0 = waitNext(f);
+	ASSERT(r0 == 0);
+	int r1 = waitNext(f);
+	ASSERT(r1 == 1);
+	int r2 = waitNext(f);
+	ASSERT(r2 == 2);
+	wait(delay(1.0));
+	int r3 = waitNext(f);
+	ASSERT(r3 == 3);
+	wait(t);
+	return Void();
+}
+
+TEST_CASE("/flow/thread/ThreadReturnPromiseStream_Error") {
+	state AsyncTaskExecutor exc(1);
+
+	{
+		ThreadReturnPromiseStream<int> s1;
+		state ThreadFutureStream<int> f1 = s1.getFuture();
+		state Future<Void> t1 = exc.post([s1 = std::move(s1)]() mutable {
+			std::this_thread::sleep_for(2s);
+			return Void();
+		});
+
+		try {
+			int _ = waitNext(f1);
+			ASSERT(false);
+		} catch (Error& e) {
+			ASSERT(e.code() == error_code_broken_promise);
+		}
+
+		wait(t1);
+	}
+
+	{
+		ThreadReturnPromiseStream<int> s2;
+		state ThreadFutureStream<int> f2 = s2.getFuture();
+		state Future<Void> t2 = exc.post([s2 = std::move(s2)]() mutable {
+			std::this_thread::sleep_for(2s);
+			s2.sendError(transaction_too_old());
+			return Void();
+		});
+
+		try {
+			int _ = waitNext(f2);
+			ASSERT(false);
+		} catch (Error& e) {
+			ASSERT(e.code() == error_code_transaction_too_old);
+		}
+
+		wait(t2);
+	}
+
+	return Void();
+}
+
+TEST_CASE("/flow/IThreadPool/ThreadReturnPromiseStream_DestroyPromise") {
+	{
+		std::cout << "ThreadReturnPromiseStream with future > 0, promise == 0, end_of_stream sent\n";
+		// After all references to PromiseStream are gone, FutureStream should still be able to get
+		// all the values if PromiseStream had reached end_of_stream.
+		state ThreadFutureStream<int> fs1 = ([]() {
+			ThreadReturnPromiseStream<int> p;
+			auto ret = p.getFuture();
+			for (int i = 0; i < 10; i++) {
+				p.send(i);
+			}
+			p.sendError(end_of_stream());
+			ASSERT(p.getFutureReferenceCount() == 1);
+			return ret;
+		})();
+
+		wait(delay(1));
+
+		state int recvd = 0;
+		try {
+			while (true) {
+				int _ = waitNext(fs1);
+				recvd += 1;
+			}
+		} catch (Error& err) {
+			if (err.code() == error_code_end_of_stream) {
+				ASSERT_EQ(recvd, 10);
+			} else {
+				ASSERT(false);
+			}
+		}
+	}
+
+	std::cout << "ThreadReturnPromiseStream with future > 0,  promise == 0, no end_of_stream\n";
+	{
+		// After all references to PromiseStream are gone, but the stream was not ended cleanly
+		// by sending end_of_stream, we should instead get broken_promise.
+		state ThreadFutureStream<int> fs2 = ([]() {
+			ThreadReturnPromiseStream<int> p;
+			auto ret = p.getFuture();
+			for (int i = 0; i < 10; i++) {
+				p.send(i);
+			}
+			ASSERT(p.getFutureReferenceCount() == 1);
+			return ret;
+		})();
+
+		wait(delay(1));
+
+		try {
+			while (true) {
+				choose {
+					when(int _ = waitNext(fs2)) {}
+					when(wait(delay(1))) {
+						ASSERT(false); // shouldn't hangup if end_of_stream is not sent.
+						break;
+					}
+				}
+			}
+		} catch (Error& err) {
+			ASSERT(err.code() == error_code_broken_promise);
+		}
+	}
+
+	return Void();
+}
```

Comment:
Nice tests.

Reply:
What happened to the test I liked that failed when we didn't copy the old backiing NotificationQueue?

Reply:
This is that test I think ("/flow/IThreadPool/ThreadReturnPromiseStream_DestroyPromise"). We also should run the kube test that originally was crashing and resulting in reverting the previous patch.

Reply:
(I can do that)